File: java/src/main/java/br/com/emendes/searchinsertposition/SearchInsertPosition.java
package br.com.emendes.searchinsertposition;

import java.util.Arrays;

/**
 * Dado um array de inteiros distintos ordenados {@code nums} e um inteiro {@code target},
 * se {@code target} for encontrado, retorne seu índice, caso contrário, retorne o índice onde estaria se fosse inserido
 * de maneira ordenada.<br>
 * Você deve escrever um algoritmo com complexidade de tempo de O(log n)<br><br>
 * Restrições:<br>
 * 1 <= nums.length <= 10⁴<br>
 * -10⁴ <= nums[i] <= 10⁴<br>
 * nums contém inteiros distintos em ordem crescente<br>
 * -10⁴ <= target <= 10⁴
 */
public class SearchInsertPosition {

  public int searchInsert(int[] nums, int target) {
    int answer = Arrays.binarySearch(nums, target);
    return answer < 0 ? (answer + 1) * (-1) : answer;
  }

}

File: javascript/problems/SearchInsertPosition.js
/**
 * Dado um array de inteiros distintos e ordenado e um inteiro target, retorne o índice do target encontrado.
 * se não, retorne o índice onde o target deveria estar.
 *
 * Você deve escrever um algortimos com complexidade de tempo O(log n).
 *
 * Restrições:
 *
 * 1 <= nums.length <= 10⁴
 *
 * -10⁴ <= nums[i] <= 10⁴
 *
 * nums contém valores distintos e ordenados em ordem crescente.
 *
 * -10⁴ <= target <= 10⁴
 *
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function (nums, target) {
  let left = 0;
  let right = nums.length - 1;
  let mid;
  while (left <= right) {
    mid = Math.floor((left + right) / 2);
    if (nums[mid] === target) return mid;
    if (nums[mid] < target) left = mid + 1;
    else right = mid - 1;
  }

  return target < nums[mid] ? mid : mid + 1;
};
